src/rust/demikernel/bindings.rs

```diff
@@ -526,6 +526,80 @@ pub extern "C" fn demi_wait_any(
     }
 }
 
+//======================================================================================================================
+// timedwait_any
+//======================================================================================================================
+
+#[no_mangle]
+pub extern "C" fn demi_timedwait_any(
+    qr_out: *mut demi_qresult_t,
+    ready_offset: *mut c_int,
+    qts: *mut demi_qtoken_t,
+    num_qts: c_int,
+    abstime: *const libc::timespec,
+) -> c_int {
+    trace!(
+        "demi_timedwait_any() {:?} {:?} {:?} {:?} {:?}",
+        qr_out,
+        ready_offset,
+        qts,
+        num_qts,
+        abstime
+    );
+
+    // Check arguments.
+    if num_qts < 0 {
+        return libc::EINVAL;
+    }
+
+    // Check for invalid timeout.
+    if abstime.is_null() {
+        warn!("abstime is a null pointer");
+        return libc::EINVAL;
+    }
+
+    // Convert timespec to SystemTime.
+    let abstime: Option<SystemTime> = {
+        if abstime.is_null() {
+            None
+        } else {
+            let timeout: Duration = Duration::from_nanos(
+                unsafe { (*abstime).tv_sec } as u64 * 1_000_000_000_ + unsafe { (*abstime).tv_nsec } as u64,
+            );
+            match SystemTime::UNIX_EPOCH.checked_add(timeout) {
+                Some(abstime) => Some(abstime),
+                None => Some(SystemTime::now()),
+            }
+        }
+    };
+
+    // Get queue tokens.
+    let qts: Vec<QToken> = {
+        let raw_qts: &[u64] = unsafe { slice::from_raw_parts(qts, num_qts as usize) };
+        raw_qts.iter().map(|i| QToken::from(*i)).collect()
+    };
+
+    // Issue operation.
+    let ret: Result<i32, Fail> = do_syscall(|libos| match libos.timedwait_any(&qts, abstime) {
+        Ok((ix, qr)) => {
+            unsafe {
+                *qr_out = qr;
+                *ready_offset = ix as c_int;
+            }
+            0
+        },
+        Err(e) => {
+            warn!("timedwait_any() failed: {:?}", e);
+            e.errno
+        },
+    });
+
+    match ret {
+        Ok(ret) => ret,
+        Err(e) => e.errno,
+    }
+}
+
```

Comment:
You could avoid a lot of code duplication here in one of two ways:

1.  Have demi_wait_any() just be a wrapper around demi_timedwait_any() that provides a special "infinite" value for the timeout.
2. Don't bother with separate calls at all, just have demi_wait_any() take a timeout value (which can be "infinite", and possibly also "immediate").

Reply:
Thanks for noting this.

I had thought about unifying these calls as well. However, I follow a software development approach where I prefer to work with incremental changes in the code base. I believe that this is important specially because some design choices will become clearer with more iterations.

Sme thoughts we still should consolidate:
- What special value should it be, so it matches a natural programming pattern for Demikernel applications?
- Iif we decide to replace `timespec` for an integer that would give us the timeout value in nanoseconds similarly to `nanosleep()`? Then, what would be that speciall value

In any case I agree that this should be reconsidered in the future. I have created an issue to keep track of this on the `wait()` and `timedwait()` system calls. https://github.com/demikernel/demikernel/issues/316

Once `timedwait_any()` is merged in the source tree I create issue for the `timedwait_any()` and `wait_any()` calls as well.

Reply:
As this is the API we expose to the user, and we had previously discussed the importance of getting the API solidified before we had too many users, I think we need to make fixing this a priority.

As for what the special values should be, I suggest we simply follow the well-established precedent of `select()`, where not specifying a timeout (via a null pointer) means "wait forever" and a timeout of zero means "poll and return immediately".

Reply:
I second the idea of prioritizing API exposure in the right way in the first/earliest attempt. But if you have a planned work item that will follow soon, it sounds good to me.